rust/datafusion/src/logical_plan/expr.rs

```diff
@@ -169,6 +170,13 @@ pub enum Expr {
     },
     /// Represents a reference to all fields in a schema.
     Wildcard,
+    /// Extract date parts (day, hour, minute) from a date / time expression
+    Extract {
```

Comment:
Alternatively, this could use `ScalarFunction` e.g. using "date_part" (which is also supported by PostgreSQL) to avoid an extra enum option, and convert the extract sql syntax to this `ScalarFunction`. I am not sure which is better?

Reply:
I have been trying to move them to the `ScalarFunction` to avoid having many variants around, mostly because adding a new item to an enum is backward incompatible and thus it may be beneficial to reserve that for operations that cannot be described by `ScalarFunction`.

Reply:
I just looked into this. The downside currently is w.r.t. performance and being able to utilize Arrow kernels. The `ScalarFunction` repeats scalar values, the date part, e.g. 'HOUR' for `date_part('HOUR', dt)` will be for repeated for each row.
In PostgreSQL, expressions are not allowed for `date_part` / `extract`, `date_trunc` etc. :

https://www.postgresql.org/docs/13/functions-datetime.html#FUNCTIONS-DATETIME-EXTRACT

> Note that here the field parameter needs to be a string value, not a name. The valid field names for date_part are the same as for extract. 

This is also happening with the existing `date_trunc` function where currently the date part (as string) is repeated / matched against for each row and also evaluated per row (see below). That won't work with the `hour` kernel for obvious reasons.

```rust
    let result = range
        .map(|i| {
            if array.is_null(i) {
                Ok(0_i64)
            } else {
                let date_time = match granularity_array.value(i) {
                    "second" => array
                        .value_as_datetime(i)
                        .and_then(|d| d.with_nanosecond(0)),
                    "minute" => array
                        .value_as_datetime(i)
                        .and_then(|d| d.with_nanosecond(0))
                        .and_then(|d| d.with_second(0)),
[...]
```
So I think here we have a few options:

* Refactor/Optimize `ScalarFunction` to also allow for scalar values, be able to check on them + support literals (I guess it should use `ColumnarValue` instead of just `Array`s).
* Have a similar (inefficient) implementation for extract from / date_part to compute as currently `date_trunc`. I think that 
* Refactor/Optimize `ScalarFunction` later and keep the Extract as is for now.

Reply:
hold my beer: #9376 :)

Reply:
🍺 awesome! 😎

Reply:
I also agree (belatedly) that having a more efficient implementation of constant function arguments (e.g. #9376) is the way to go!

In terms of adding variants to `Expr` I think it will need be done when the semantics of whatever expression is being added can't realistically be expressed as a function (e.g. CASE). 

So in this case, given @jorgecarleitao is cranking along with #9376 it seems like perhaps this PR should perhaps try and translate `EXTRACT` into a function.

Reply:
Yes, I think it is worth it to wait for the other PR to land and convert it to use the scalar function. Then we could also add the function alias `date_part` easily next to the `extract` syntax, which are both supported by PostgreSQL.